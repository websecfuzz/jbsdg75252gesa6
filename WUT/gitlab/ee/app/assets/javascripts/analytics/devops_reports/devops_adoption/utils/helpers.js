import { sprintf } from '~/locale';
import { joinPaths } from '~/lib/utils/url_utility';
import { GROUP_DEVOPS_PATH } from '../constants';

/**
 * A helper function which accepts the enabledNamespaces,
 *
 * @param {Object} params the enabledNamespaces data, and the check for open modals
 *
 * @return {Boolean} a boolean to determine if table data should be polled
 */
export const shouldPollTableData = ({ enabledNamespaces, openModal }) => {
  if (openModal) {
    return false;
  }
  if (!enabledNamespaces.length) {
    return true;
  }

  const anyPendingEnabledNamespaces = enabledNamespaces.some(
    (node) => node.latestSnapshot === null,
  );

  return anyPendingEnabledNamespaces;
};

/**
 * A helper function which extracts the total feature adoption count for a group
 * of snapshot data, filtered out by specific features / columns
 *
 * @param { Array } snapshots the snapshot data for a given group node
 * @param { Array } cols the columns which need to be used for the calculation
 *
 * @return { Array } an array containing the adopted counts for the given columns
 */
export const getAdoptedCountsByCols = (snapshots, cols) => {
  return snapshots.reduce((acc, snapshot) => {
    const adoptedCount = cols.reduce((adopted, col) => {
      return snapshot[col.key] ? adopted + 1 : adopted;
    }, 0);

    return [...acc, adoptedCount];
  }, []);
};

/**
 * A helper function which computes the DevOps Adoption feature path
 * given a specific group path
 *
 * @param { String } fullPath the full path for the group
 *
 * @return { String } the path for the group level DevOps Adoption feature
 */
export const getGroupAdoptionPath = (fullPath) => {
  const prefix = gon.relative_url_root || '';
  return fullPath ? joinPaths(prefix, sprintf(GROUP_DEVOPS_PATH, { fullPath })) : null;
};
