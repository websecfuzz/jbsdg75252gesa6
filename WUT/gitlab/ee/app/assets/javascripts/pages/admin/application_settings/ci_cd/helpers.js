/**
 * Filters [items] based on a given [searchTerm].
 * Catagories with no items after filtering are not included in the returned object.
 * @param {Object} allItems - { <categoryName>: [{ name, id }] }
 * @param {String} searchTerm
 * @returns {Object}
 */
export function filterItems(allItems, searchTerm) {
  return Object.entries(allItems)
    .map(([key, items]) => ({
      text: key,
      options: items
        .filter((item) => item.name.toLowerCase().includes(searchTerm))
        .map((item) => ({
          text: item.name,
          value: item.key,
        })),
    }))
    .filter((group) => group.options.length > 0);
}
