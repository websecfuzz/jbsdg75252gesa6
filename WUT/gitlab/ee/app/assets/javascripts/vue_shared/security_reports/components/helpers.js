import { __ } from '~/locale';
import { EMPTY_BODY_MESSAGE } from './constants';

/**
 * A helper function which validates the passed
 * in body string.
 *
 * It returns an empty string if the body has explicitly
 * been passed in as an empty string, a fallback
 * message if the body is null / undefined, else
 * it will return the original body string.
 *
 * @param {String} body the body message
 *
 * @return {String} the validated body message
 */
export const bodyWithFallBack = (body) => (body === '' ? '' : body || EMPTY_BODY_MESSAGE);

/**
 * Returns a string representation of the HTTP request
 *
 * @param {Object} httpData
 * @param {String} httpData.body
 * @param {String} httpData.method
 * @param {String} httpData.url
 * @param {{name: String, value: String}[]} httpData.headers
 *
 * @returns {String}
 */
export const getHttpString = (httpData) => {
  if (!httpData) {
    return '';
  }

  const { body, headers, method, url, statusCode, reasonPhrase } = httpData;
  const methodOrStatusCode = method || statusCode;
  const urlOrReasonPhrase = url || reasonPhrase;
  const headerString = headers.map(({ name, value }) => `${name}: ${value}`).join('\n');

  return `${methodOrStatusCode} ${urlOrReasonPhrase}\n${headerString}\n\n${bodyWithFallBack(body)}`;
};

export const getDismissalNoteEventText = ({ hasProject, hasPipeline, hasDismissalReason }) => {
  if (hasDismissalReason) {
    if (hasPipeline && hasProject) {
      return __(
        '%{statusStart}Dismissed%{statusEnd}: %{dismissalReason} on pipeline %{pipelineLink} at %{projectLink}',
      );
    }

    if (hasPipeline) {
      return __(
        '%{statusStart}Dismissed%{statusEnd}: %{dismissalReason} on pipeline %{pipelineLink}',
      );
    }

    if (hasProject) {
      return __('%{statusStart}Dismissed%{statusEnd}: %{dismissalReason} at %{projectLink}');
    }

    return __('%{statusStart}Dismissed%{statusEnd}: %{dismissalReason}');
  }

  if (hasPipeline && hasProject) {
    return __('%{statusStart}Dismissed%{statusEnd} on pipeline %{pipelineLink} at %{projectLink}');
  }

  if (hasPipeline) {
    return __('%{statusStart}Dismissed%{statusEnd} on pipeline %{pipelineLink}');
  }

  if (hasProject) {
    return __('%{statusStart}Dismissed%{statusEnd} at %{projectLink}');
  }

  return __('%{statusStart}Dismissed%{statusEnd}');
};
