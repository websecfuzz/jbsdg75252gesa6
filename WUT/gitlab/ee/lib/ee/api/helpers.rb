# frozen_string_literal: true

module EE
  module API
    module Helpers
      extend ::Gitlab::Utils::Override
      include ::Gitlab::Utils::StrongMemoize

      def require_node_to_be_enabled!
        forbidden! 'Geo node is disabled.' unless ::Gitlab::Geo.current_node&.enabled?
      end

      def gitlab_geo_node_token?
        headers['Authorization']&.start_with?(::Gitlab::Geo::BaseRequest::GITLAB_GEO_AUTH_TOKEN_TYPE)
      end

      def authenticate_by_gitlab_geo_node_token!
        unauthorized! unless authorization_header_valid?
      rescue ::Gitlab::Geo::InvalidDecryptionKeyError, ::Gitlab::Geo::InvalidSignatureTimeError => e
        render_api_error!(e.to_s, 401)
      end

      def geo_jwt_decoder
        return unless gitlab_geo_node_token?

        strong_memoize(:geo_jwt_decoder) do
          ::Gitlab::Geo::JwtRequestDecoder.new(headers['Authorization'])
        end
      end

      # Update the jwt_decoder to allow authorization of disabled (paused) nodes
      def allow_paused_nodes!
        geo_jwt_decoder.include_disabled!
      end

      def check_gitlab_geo_request_ip!
        unauthorized! unless ::Gitlab::Geo.allowed_ip?(request.ip)
      end

      def authorization_header_valid?
        return unless gitlab_geo_node_token?

        scope = geo_jwt_decoder.decode.try { |x| x[:scope] }
        scope == ::Gitlab::Geo::API_SCOPE
      end

      def check_project_feature_available!(feature)
        not_found! unless user_project.feature_available?(feature)
      end

      def authorize_change_param(subject, *keys)
        keys.each do |key|
          authorize!(:"change_#{key}", subject) if params.has_key?(key)
        end
      end

      # Normally, only admin users should have access to see LDAP
      # groups. However, due to the "Allow group owners to manage LDAP-related
      # group settings" setting, any group owner can sync LDAP groups with
      # their project.
      #
      # In the future, we should also check that the user has access to manage
      # a specific group so that we can use the Ability class.
      def authenticated_with_ldap_admin_access!
        authenticate!

        forbidden! unless current_user.admin? ||
          ::Gitlab::CurrentSettings.current_application_settings
            .allow_group_owners_to_manage_ldap
      end

      override :read_project_ability
      def read_project_ability
        # CI job token authentication:
        # this method grants limited privileged for admin users
        # admin users can only access project if they are direct member
        job_token_authentication? ? :build_read_project : super
      end

      override :find_group!
      def find_group!(id, organization: nil)
        # CI job token authentication:
        # currently we do not allow any group access for CI job token
        if job_token_authentication?
          not_found!('Group')
        else
          super
        end
      end

      override :find_pipeline!
      def find_pipeline!(id)
        if job_token_authentication?
          not_found!('Pipeline')
        else
          super
        end
      end

      # rubocop: disable CodeReuse/ActiveRecord
      def find_group_epic(iid)
        EpicsFinder.new(current_user, group_id: user_group.id).find_by!(iid: iid)
      end
      # rubocop: enable CodeReuse/ActiveRecord

      def find_or_create_subscription_add_on!(name, namespace = nil)
        not_found!('Subscription Add-on') unless GitlabSubscriptions::AddOn.names[name]

        GitlabSubscriptions::AddOn.find_or_create_by_name(name, namespace)
      end

      # rubocop: disable CodeReuse/ActiveRecord
      def find_subscription_add_on_purchase!(namespace, add_on)
        GitlabSubscriptions::AddOnPurchase.find_by!(
          namespace: namespace,
          add_on: add_on
        )
      end
      # rubocop: enable CodeReuse/ActiveRecord

      def user_pipeline
        @pipeline ||= find_pipeline!(params[:id])
      end

      private

      override :project_finder_params_ee
      def project_finder_params_ee
        finder_params = {}

        finder_params[:include_hidden] = declared_params[:include_hidden] if declared_params[:include_hidden]
        finder_params[:with_security_reports] = true if params[:with_security_reports].present?

        finder_params
      end

      override :send_git_archive
      def send_git_archive(repository, **kwargs)
        result = ::Users::Abuse::ProjectsDownloadBanCheckService.execute(current_user, repository.project)
        forbidden!(_('You are not allowed to download code from this project.')) if result.error?

        project = repository.project
        audit_event_type = project.public? ? 'public_repository_download_operation' : 'repository_download_operation'
        audit_context = {
          name: audit_event_type,
          ip_address: ip_address,
          author: current_user || ::Gitlab::Audit::UnauthenticatedAuthor.new,
          target: project,
          scope: project,
          message: "Repository Download Started",
          target_details: project.full_path
        }
        ::Gitlab::Audit::Auditor.audit(audit_context)

        super
      end

      def private_token
        params[::APIGuard::PRIVATE_TOKEN_PARAM] || env[::APIGuard::PRIVATE_TOKEN_HEADER]
      end

      def warden
        env['warden']
      end

      # Check if the request is GET/HEAD, or if CSRF token is valid.
      def verified_request?
        ::Gitlab::RequestForgeryProtection.verified?(env)
      end

      # Check the Rails session for valid authentication details
      def find_user_from_warden
        warden.try(:authenticate) if verified_request?
      end

      def geo_token
        ::Gitlab::Geo.current_node.system_hook.token
      end

      def authorize_manage_saml!(group)
        unauthorized! unless can?(current_user, :admin_group_saml, group)
      end

      def check_group_saml_configured
        forbidden!('Group SAML not enabled.') unless ::Gitlab::Auth::GroupSaml::Config.enabled?
      end

      def check_instance_saml_configured
        forbidden!('SAML not enabled.') unless ::Gitlab::Auth::Saml::Config.enabled?
      end
    end
  end
end
