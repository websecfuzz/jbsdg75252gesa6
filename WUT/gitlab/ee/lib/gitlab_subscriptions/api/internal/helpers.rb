# frozen_string_literal: true

module GitlabSubscriptions
  module API
    module Internal
      module Helpers
        def jwt_request?
          headers[Auth::INTERNAL_API_REQUEST_HEADER].present?
        end

        def authenticate_from_jwt!
          unauthorized! unless Auth.verify_api_request(headers)
        end
      end
    end
  end
end
