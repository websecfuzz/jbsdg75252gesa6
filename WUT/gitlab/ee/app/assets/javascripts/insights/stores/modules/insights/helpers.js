import { pairDataAndLabels, buildNullSeries } from 'ee/analytics/shared/utils';
import { BASE_NULL_SERIES_OPTIONS, BASE_SERIES_DATA_OPTIONS } from 'ee/analytics/shared/constants';
import {
  CHART_TYPES,
  INSIGHTS_DATA_SOURCE_DORA,
  INSIGHTS_NO_DATA_TOOLTIP,
} from 'ee/insights/constants';
import { __, s__ } from '~/locale';

const getGroupByValue = (queryParams) => {
  return queryParams.group_by || queryParams.params.group_by;
};

const getTypeValue = (queryParams) => {
  return queryParams.issuable_type || queryParams.params.issuable_type || queryParams.params.metric;
};

const getAxisTitle = (label) => {
  switch (label) {
    case 'day':
      return __('Days');
    case 'week':
      return __('Weeks');
    case 'month':
      return __('Months');
    case 'issue':
      return __('Issues');
    case 'merge_request':
      return __('Merge requests');
    case 'deployment_frequency':
      return s__('DORA4Metrics|Deployment frequency');
    case 'lead_time_for_changes':
      return s__('DORA4Metrics|Lead time for changes (median days)');
    case 'time_to_restore_service':
      return s__('DORA4Metrics|Time to restore service (median days)');
    case 'change_failure_rate':
      return s__('DORA4Metrics|Change failure rate (percentage)');
    default:
      return '';
  }
};

export const transformChartDataForGlCharts = (
  { type, query: queryParams },
  { labels, datasets },
) => {
  const yAxisTitle = getAxisTitle(getTypeValue(queryParams));
  const formattedData = {
    xAxisTitle: getAxisTitle(getGroupByValue(queryParams)),
    yAxisTitle,
    labels,
    datasets: [],
    seriesNames: [],
  };

  const { data_source: dataSource = '' } = queryParams;

  switch (type) {
    case CHART_TYPES.BAR:
      formattedData.datasets = [
        {
          name: 'all',
          data: labels.map((label, i) => [label, datasets[0].data[i]]),
        },
      ];
      break;
    case CHART_TYPES.STACKED_BAR:
      formattedData.datasets.push(
        ...datasets.map((dataset) => ({
          name: dataset.label,
          data: dataset.data,
        })),
      );
      break;
    case CHART_TYPES.LINE:
      if (dataSource === INSIGHTS_DATA_SOURCE_DORA) {
        const paired = pairDataAndLabels({
          datasetNames: [yAxisTitle],
          axisLabels: labels,
          datasets,
        });

        formattedData.datasets = buildNullSeries({
          seriesData: paired,
          seriesDataOptions: BASE_SERIES_DATA_OPTIONS,
          nullSeriesTitle: INSIGHTS_NO_DATA_TOOLTIP,
          nullSeriesOptions: BASE_NULL_SERIES_OPTIONS,
        });
      } else {
        formattedData.datasets.push(
          ...datasets.map((dataset) => ({
            name: dataset.label,
            data: labels.map((label, i) => [label, dataset.data[i]]),
          })),
        );
      }
      break;
    default:
      formattedData.datasets = { all: labels.map((label, i) => [label, datasets[0].data[i]]) };
  }
  return formattedData;
};

export default {
  transformChartDataForGlCharts,
};
