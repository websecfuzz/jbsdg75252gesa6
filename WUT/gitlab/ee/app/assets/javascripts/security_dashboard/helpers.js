import isPlainObject from 'lodash/isPlainObject';
import { REPORT_TYPES_ALL } from 'ee/security_dashboard/constants';
import { convertObjectPropsToSnakeCase } from '~/lib/utils/common_utils';
import { __ } from '~/locale';

/**
 * Provided a security reports summary from the GraphQL API, this returns an array of arrays
 * representing a properly formatted report ready to be displayed in the UI. Each sub-array consists
 * of the user-friend report's name, and the summary's payload. Note that summary entries are
 * considered empty and are filtered out of the return if the payload is `null` or don't include
 * a vulnerabilitiesCount property. Report types whose name can't be matched to a user-friendly
 * name are filtered out as well.
 *
 * Take the following summary for example:
 * {
 *   containerScanning: { vulnerabilitiesCount: 123 },
 *   invalidReportType: { vulnerabilitiesCount: 123 },
 *   dast: null,
 * }
 *
 * The formatted summary would look like this:
 * [
 *   ['containerScanning', { vulnerabilitiesCount: 123 }]
 * ]
 *
 * Note that `invalidReportType` was filtered out as it can't be matched with a user-friendly name,
 * and the DAST report was omitted because it's empty (`null`).
 *
 * @param {Object} rawSummary
 * @returns {Array}
 */
export const getFormattedSummary = (rawSummary = {}) => {
  if (!isPlainObject(rawSummary)) {
    return [];
  }
  // Convert keys to snake case so they can be matched against REPORT_TYPES keys for translation
  const snakeCasedSummary = convertObjectPropsToSnakeCase(rawSummary);
  // Convert object to an array of entries to make it easier to loop through
  const summaryEntries = Object.entries(snakeCasedSummary);
  // Filter out empty entries as we don't want to display those in the summary
  const withoutEmptyEntries = summaryEntries.filter(
    ([, scanSummary]) => scanSummary?.vulnerabilitiesCount !== undefined,
  );
  // Replace keys with translations found in REPORT_TYPES if available
  const formattedEntries = withoutEmptyEntries.map(([scanType, scanSummary]) => {
    const name = REPORT_TYPES_ALL[scanType];
    return name ? [name, scanSummary] : null;
  });
  // Filter out keys that could not be matched with any translation and are thus considered invalid
  return formattedEntries.filter((entry) => entry !== null);
};

/**
 * Limits the number of projects displayed per vulnerability grade.
 *
 * Takes an array of vulnerability grade data and ensures that each grade
 * shows at most the specified number of projects. Grades with fewer projects
 * than the limit are returned unchanged.
 *
 * @param {Array} vulnerabilityGradesQueryResults - Array of vulnerability grade objects from GraphQL query
 * @param {number} [maxProjects=5] - Maximum number of projects to show per grade
 * @returns {Array} Array of vulnerability grade objects with limited project nodes
 */
export const limitVulnerabilityGradeProjects = (
  vulnerabilityGradesQueryResults,
  maxProjects = 5,
) => {
  return vulnerabilityGradesQueryResults.map((gradeData) => {
    const nodes = gradeData.projects?.nodes;

    if (!nodes || nodes.length <= maxProjects) {
      return gradeData;
    }

    return {
      ...gradeData,
      projects: {
        ...gradeData.projects,
        nodes: nodes.slice(0, maxProjects),
      },
    };
  });
};

export const PROJECT_LOADING_ERROR_MESSAGE = __('An error occurred while retrieving projects.');

/**
 * Custom error class for PDF export operations that are not ready.
 *
 * @param {string} message - User-friendly error message describing why the export isn't ready
 */
export class PdfExportError extends Error {
  constructor(message) {
    super(message);
    this.name = 'PdfExportError';
  }
}

export default () => ({});
