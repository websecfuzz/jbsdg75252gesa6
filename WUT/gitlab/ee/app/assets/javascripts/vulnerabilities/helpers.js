import { getIdFromGraphQLId } from '~/graphql_shared/utils';
import { isAbsolute, isValidURL } from '~/lib/utils/url_utility';
import {
  REGEXES,
  SUPPORTED_IDENTIFIER_TYPE_CWE,
  SUPPORTED_IDENTIFIER_TYPE_OWASP,
  VULNERABILITY_TAB_INDEX_TO_NAME,
} from './constants';

// Get the issue in the format expected by the descendant components of related_issues_block.vue.
export const getFormattedIssue = (issue) => ({
  ...issue,
  reference: `#${issue.iid}`,
  path: issue.web_url,
});

export const getAddRelatedIssueRequestParams = (reference, defaultProjectId) => {
  let issueId = reference;
  let projectId = defaultProjectId;

  // If the reference is an issue number, parse out just the issue number.
  if (REGEXES.ISSUE_FORMAT.test(reference)) {
    [, issueId] = REGEXES.ISSUE_FORMAT.exec(reference);
  }
  // If the reference is an absolute URL and matches the issues URL format, parse out the project and issue.
  else if (isValidURL(reference) && isAbsolute(reference)) {
    const { pathname } = new URL(reference);

    if (REGEXES.LINK_FORMAT.test(pathname)) {
      [, projectId, issueId] = REGEXES.LINK_FORMAT.exec(pathname);
    }
  }

  return { target_issue_iid: issueId, target_project_id: projectId };
};

export const normalizeGraphQLNote = (note) => {
  if (!note) {
    return null;
  }

  return {
    ...note,
    id: getIdFromGraphQLId(note.id),
    author: {
      ...note.author,
      id: getIdFromGraphQLId(note.author.id),
      path: note.author.webPath,
    },
  };
};

export const normalizeGraphQLVulnerability = (vulnerability) => {
  if (!vulnerability) {
    return null;
  }

  const newVulnerability = { ...vulnerability };

  if (vulnerability.id) {
    newVulnerability.id = getIdFromGraphQLId(vulnerability.id);
  }

  if (vulnerability.state) {
    newVulnerability.state = vulnerability.state.toLowerCase();
  }

  ['confirmed', 'resolved', 'dismissed'].forEach((state) => {
    if (vulnerability[`${state}By`]?.id) {
      newVulnerability[`${state}ById`] = getIdFromGraphQLId(vulnerability[`${state}By`].id);
      delete newVulnerability[`${state}By`];
    }
  });

  return newVulnerability;
};

export const normalizeGraphQLLastStateTransition = (graphQLVulnerability, vulnerability) => {
  const stateTransitions = [...vulnerability.stateTransitions];

  // The vulnerability status mutation only returns 1 stateTransition
  const [graphQLLastStateTransitions] = graphQLVulnerability.stateTransitions.nodes;
  stateTransitions.push({
    ...graphQLLastStateTransitions,
    dismissalReason: graphQLLastStateTransitions.dismissalReason?.toLowerCase(),
  });

  return { stateTransitions };
};

export const formatIdentifierExternalIds = ({ externalType, externalId, name }) => {
  return `[${externalType}]-[${externalId}]-[${name}]`;
};

export const isSupportedIdentifier = (externalType) => {
  return (
    externalType?.toLowerCase() === SUPPORTED_IDENTIFIER_TYPE_CWE ||
    // Case matters here. owasp and OWASP require different configuration
    // Currently, our API only support lowercase owasp
    // Uppercase OWASP will be supported in a follow up issue:
    // https://gitlab.com/gitlab-org/gitlab/-/issues/366556
    externalType === SUPPORTED_IDENTIFIER_TYPE_OWASP
  );
};

export const getRefFromBlobPath = (path) => {
  // Matches the 40-character hex string of the Git SHA in the blob path, for example:
  // "/group/project/-/blob/cdeda7ae724a332e008d17245209d5edd9ba6499/src/file.js"
  // will match "cdeda7ae724a332e008d17245209d5edd9ba6499".
  if (typeof path === 'string') {
    const match = path.match(/blob\/([a-f0-9]{40})/);
    return match ? match[1] : '';
  }
  return '';
};

/**
 *  Determines the tab index for code flow based on the route query parameters.
 */
export const getTabIndexForCodeFlowPage = (route) => {
  if (route?.query?.tab) {
    return 1;
  }
  return 0;
};

/**
 * @typedef {Object} TabNavigationParams
 * @property {string} path - The current route path.
 * @property {number} index - The index of the tab to be set.
 * @property {number} [selectedIndex] - The currently selected tab index. This is optional as it might be undefined.
 */

/**
 * Updates the route to reflect the selected tab index.
 * This function is used to update the URL when switching between different views
 * of the vulnerability, between 'details' and 'code flow'
 * @param {Object} router - The router object used for navigation.
 * @param {TabNavigationParams} routeProps - The route properties including path and tab indices.
 */
export const setTabIndexForCodeFlowPage = (router, routeProps) => {
  const { path, index, selectedIndex } = routeProps;

  // Don't set the tab index if it's the same. This prevents some querystring bugs.
  if (index === selectedIndex) {
    return;
  }
  router.push({
    path,
    query: { tab: VULNERABILITY_TAB_INDEX_TO_NAME[index] },
  });
};
