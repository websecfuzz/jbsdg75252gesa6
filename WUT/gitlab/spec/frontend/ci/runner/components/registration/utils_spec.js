import { TEST_HOST } from 'helpers/test_constants';
import {
  DEFAULT_PLATFORM,
  LINUX_PLATFORM,
  MACOS_PLATFORM,
  WINDOWS_PLATFORM,
} from '~/ci/runner/constants';

import {
  commandPrompt,
  registerCommand,
  runCommand,
  installScript,
  platformArchitectures,
} from '~/ci/runner/components/registration/utils';

import { mockAuthenticationToken } from '../../mock_data';

describe('registration utils', () => {
  beforeEach(() => {
    window.gon.gitlab_url = TEST_HOST;
  });

  describe.each([LINUX_PLATFORM, MACOS_PLATFORM, WINDOWS_PLATFORM])(
    'for "%s" platform',
    (platform) => {
      it('commandPrompt is correct', () => {
        expect(commandPrompt({ platform })).toMatchSnapshot();
      });

      it('registerCommand is correct', () => {
        expect(
          registerCommand({
            platform,
            token: mockAuthenticationToken,
          }),
        ).toMatchSnapshot();

        expect(registerCommand({ platform })).toMatchSnapshot();
      });

      it('runCommand is correct', () => {
        expect(runCommand({ platform })).toMatchSnapshot();
      });
    },
  );

  describe('for missing platform', () => {
    it('commandPrompt uses the default', () => {
      const expected = commandPrompt({ platform: DEFAULT_PLATFORM });

      expect(commandPrompt({ platform: null })).toEqual(expected);
      expect(commandPrompt({ platform: undefined })).toEqual(expected);
    });

    it('registerCommand uses the default', () => {
      const expected = registerCommand({
        platform: DEFAULT_PLATFORM,
        token: mockAuthenticationToken,
      });

      expect(registerCommand({ platform: null, token: mockAuthenticationToken })).toEqual(expected);
      expect(registerCommand({ platform: undefined, token: mockAuthenticationToken })).toEqual(
        expected,
      );
    });

    it('runCommand uses the default', () => {
      const expected = runCommand({ platform: DEFAULT_PLATFORM });

      expect(runCommand({ platform: null })).toEqual(expected);
      expect(runCommand({ platform: undefined })).toEqual(expected);
    });
  });

  describe.each([LINUX_PLATFORM, MACOS_PLATFORM, WINDOWS_PLATFORM])(
    'for "%s" platform',
    (platform) => {
      describe('platformArchitectures', () => {
        it('returns correct list of architectures', () => {
          expect(platformArchitectures({ platform })).toMatchSnapshot();
        });
      });

      describe('installScript', () => {
        const architectures = platformArchitectures({ platform });

        it.each(architectures)('is correct for "%s" architecture', (architecture) => {
          expect(installScript({ platform, architecture })).toMatchSnapshot();
        });
      });
    },
  );
});
