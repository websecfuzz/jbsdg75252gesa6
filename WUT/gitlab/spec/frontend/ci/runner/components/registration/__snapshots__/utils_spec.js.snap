// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`registration utils for "linux" platform commandPrompt is correct 1`] = `"$"`;

exports[`registration utils for "linux" platform installScript is correct for "386" architecture 1`] = `
"# Download the binary for your system
sudo curl -L --output /usr/local/bin/gitlab-runner https://gitlab-runner-downloads.s3.amazonaws.com/latest/binaries/gitlab-runner-linux-386

# Give it permission to execute
sudo chmod +x /usr/local/bin/gitlab-runner

# Create a GitLab Runner user
sudo useradd --comment 'GitLab Runner' --create-home gitlab-runner --shell /bin/bash

# Install and run as a service
sudo gitlab-runner install --user=gitlab-runner --working-directory=/home/gitlab-runner
sudo gitlab-runner start"
`;

exports[`registration utils for "linux" platform installScript is correct for "amd64" architecture 1`] = `
"# Download the binary for your system
sudo curl -L --output /usr/local/bin/gitlab-runner https://gitlab-runner-downloads.s3.amazonaws.com/latest/binaries/gitlab-runner-linux-amd64

# Give it permission to execute
sudo chmod +x /usr/local/bin/gitlab-runner

# Create a GitLab Runner user
sudo useradd --comment 'GitLab Runner' --create-home gitlab-runner --shell /bin/bash

# Install and run as a service
sudo gitlab-runner install --user=gitlab-runner --working-directory=/home/gitlab-runner
sudo gitlab-runner start"
`;

exports[`registration utils for "linux" platform installScript is correct for "arm" architecture 1`] = `
"# Download the binary for your system
sudo curl -L --output /usr/local/bin/gitlab-runner https://gitlab-runner-downloads.s3.amazonaws.com/latest/binaries/gitlab-runner-linux-arm

# Give it permission to execute
sudo chmod +x /usr/local/bin/gitlab-runner

# Create a GitLab Runner user
sudo useradd --comment 'GitLab Runner' --create-home gitlab-runner --shell /bin/bash

# Install and run as a service
sudo gitlab-runner install --user=gitlab-runner --working-directory=/home/gitlab-runner
sudo gitlab-runner start"
`;

exports[`registration utils for "linux" platform installScript is correct for "arm64" architecture 1`] = `
"# Download the binary for your system
sudo curl -L --output /usr/local/bin/gitlab-runner https://gitlab-runner-downloads.s3.amazonaws.com/latest/binaries/gitlab-runner-linux-arm64

# Give it permission to execute
sudo chmod +x /usr/local/bin/gitlab-runner

# Create a GitLab Runner user
sudo useradd --comment 'GitLab Runner' --create-home gitlab-runner --shell /bin/bash

# Install and run as a service
sudo gitlab-runner install --user=gitlab-runner --working-directory=/home/gitlab-runner
sudo gitlab-runner start"
`;

exports[`registration utils for "linux" platform platformArchitectures returns correct list of architectures 1`] = `
[
  "amd64",
  "386",
  "arm",
  "arm64",
]
`;

exports[`registration utils for "linux" platform registerCommand is correct 1`] = `
[
  "gitlab-runner register",
  "  --url http://test.host",
  "  --token MOCK_AUTHENTICATION_TOKEN",
]
`;

exports[`registration utils for "linux" platform registerCommand is correct 2`] = `
[
  "gitlab-runner register",
  "  --url http://test.host",
]
`;

exports[`registration utils for "linux" platform runCommand is correct 1`] = `"gitlab-runner run"`;

exports[`registration utils for "osx" platform commandPrompt is correct 1`] = `"$"`;

exports[`registration utils for "osx" platform installScript is correct for "amd64" architecture 1`] = `
"# Download the binary for your system
sudo curl --output /usr/local/bin/gitlab-runner https://gitlab-runner-downloads.s3.amazonaws.com/latest/binaries/gitlab-runner-darwin-amd64

# Give it permission to execute
sudo chmod +x /usr/local/bin/gitlab-runner

# The rest of the commands execute as the user who will run the runner
# Register the runner (steps below), then run
cd ~
gitlab-runner install
gitlab-runner start"
`;

exports[`registration utils for "osx" platform installScript is correct for "arm64" architecture 1`] = `
"# Download the binary for your system
sudo curl --output /usr/local/bin/gitlab-runner https://gitlab-runner-downloads.s3.amazonaws.com/latest/binaries/gitlab-runner-darwin-arm64

# Give it permission to execute
sudo chmod +x /usr/local/bin/gitlab-runner

# The rest of the commands execute as the user who will run the runner
# Register the runner (steps below), then run
cd ~
gitlab-runner install
gitlab-runner start"
`;

exports[`registration utils for "osx" platform platformArchitectures returns correct list of architectures 1`] = `
[
  "amd64",
  "arm64",
]
`;

exports[`registration utils for "osx" platform registerCommand is correct 1`] = `
[
  "gitlab-runner register",
  "  --url http://test.host",
  "  --token MOCK_AUTHENTICATION_TOKEN",
]
`;

exports[`registration utils for "osx" platform registerCommand is correct 2`] = `
[
  "gitlab-runner register",
  "  --url http://test.host",
]
`;

exports[`registration utils for "osx" platform runCommand is correct 1`] = `"gitlab-runner run"`;

exports[`registration utils for "windows" platform commandPrompt is correct 1`] = `">"`;

exports[`registration utils for "windows" platform installScript is correct for "386" architecture 1`] = `
"# Run PowerShell: https://docs.microsoft.com/en-us/powershell/scripting/windows-powershell/starting-windows-powershell?view=powershell-7#with-administrative-privileges-run-as-administrator
# Create a folder somewhere on your system, for example: C:\\GitLab-Runner
New-Item -Path 'C:\\GitLab-Runner' -ItemType Directory

# Change to the folder
cd 'C:\\GitLab-Runner'

# Download binary
Invoke-WebRequest -Uri "https://gitlab-runner-downloads.s3.amazonaws.com/latest/binaries/gitlab-runner-windows-386.exe" -OutFile "gitlab-runner.exe"

# Register the runner (steps below), then run
.\\gitlab-runner.exe install
.\\gitlab-runner.exe start"
`;

exports[`registration utils for "windows" platform installScript is correct for "amd64" architecture 1`] = `
"# Run PowerShell: https://docs.microsoft.com/en-us/powershell/scripting/windows-powershell/starting-windows-powershell?view=powershell-7#with-administrative-privileges-run-as-administrator
# Create a folder somewhere on your system, for example: C:\\GitLab-Runner
New-Item -Path 'C:\\GitLab-Runner' -ItemType Directory

# Change to the folder
cd 'C:\\GitLab-Runner'

# Download binary
Invoke-WebRequest -Uri "https://gitlab-runner-downloads.s3.amazonaws.com/latest/binaries/gitlab-runner-windows-amd64.exe" -OutFile "gitlab-runner.exe"

# Register the runner (steps below), then run
.\\gitlab-runner.exe install
.\\gitlab-runner.exe start"
`;

exports[`registration utils for "windows" platform platformArchitectures returns correct list of architectures 1`] = `
[
  "amd64",
  "386",
]
`;

exports[`registration utils for "windows" platform registerCommand is correct 1`] = `
[
  ".\\gitlab-runner.exe register",
  "  --url http://test.host",
  "  --token MOCK_AUTHENTICATION_TOKEN",
]
`;

exports[`registration utils for "windows" platform registerCommand is correct 2`] = `
[
  ".\\gitlab-runner.exe register",
  "  --url http://test.host",
]
`;

exports[`registration utils for "windows" platform runCommand is correct 1`] = `".\\gitlab-runner.exe run"`;
